File: 0016-3sum-closest/0016-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int ans_sum = 0;
        int min_diff = INT_MAX;
        int n = nums.size();

        sort(nums.begin(),nums.end());
        
        for (int i = 0; i < n - 2; i++) {
            int l = i + 1;
            int r = n - 1;

            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                int diff = abs(sum - target);
                if (sum == target)
                    return sum;
                if (sum < target) {
                    if (diff < min_diff) {
                        min_diff = diff;
                        ans_sum = sum;
                    }
                    l++;

                } else {
                    if (diff < min_diff) {
                        min_diff = diff;
                        ans_sum = sum;
                    }
                    r--;
                }
            }
        }
        return ans_sum;
    }
};

File: problems/0016-3sum-closest/0016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function (nums, target) {
    nums.sort((a, b) => a - b);

    let min_diff = Number.MAX_SAFE_INTEGER;
    let closestSum = 0;

    let n = nums.length;
    for (let i = 0; i < n - 1; i++) {
        let left = i + 1,
            right = n - 1;

        while (left < right) {
            let curSum = nums[i] + nums[left] + nums[right];

            if (curSum === target) return curSum;

            let diff = Math.abs(curSum - target);
            if (diff < min_diff) {
                min_diff = diff;
                closestSum = curSum;
            }

            if (curSum > target) right--;
            else left++;
        }
    }

    return closestSum;
};
